### ISSUE TITLE
Model Doesnt Converge

### ISSUE THREAD

**Issue by mateenahmedai** [mhagglun/Speech-Recognition#1]
Hi! I am training the model on colab and have installed librosa, tensorflow_io and the 'requirements.txt' file as per required dependencies.

1) the GPU doesn't get detected on colab, so i am using a CPU instance
2) After training on CPU, the loss doesn't converge and accuracy stays at 3% during the whole time.
3) All test files during the testing are recognized as 'off' utterance.

Upon closer inspection, i found out that scipy is unable to read some of the 'background_noises', therefore i converted the 'doing_the_dishes' and other background noise files to mp3 and back into .wav format, doing so solved the problem, but still the accuracy stays at 3-4% during the whole training.

Error Received During the Scipy .wav file reading:
WavFileWarning: Chunk (non-data) not understood, skipping it.  (rate, signal) = wav.read(source_file)                # Read audiofile

I also tried running the code on local machine but suffered the same fate, attached below is the output from local machine...

Epoch 1/20

2021-09-07 05:06:08.934147: I tensorflow/core/profiler/lib/profiler_session.cc:131] Profiler session initializing.
2021-09-07 05:06:08.934176: I tensorflow/core/profiler/lib/profiler_session.cc:146] Profiler session started.
2021-09-07 05:06:08.934227: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1614] Profiler found 1 GPUs
2021-09-07 05:06:08.935781: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcupti.so.11.2'; dlerror: libcupti.so.11.2: cannot open shared object file: No such file or directory
2021-09-07 05:06:08.936044: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcupti.so'; dlerror: libcupti.so: cannot open shared object file: No such file or directory
2021-09-07 05:06:08.936059: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1666] function cupti_interface_->Subscribe( &subscriber_, (CUpti_CallbackFunc)ApiCallback, this)failed with error CUPTI could not be loaded or symbol could not be found.
2021-09-07 05:06:08.936090: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session tear down.
2021-09-07 05:06:08.936103: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1757] function cupti_interface_->Finalize()failed with error CUPTI could not be loaded or symbol could not be found.

  1/753 [..............................] - ETA: 15:11 - loss: 2.9148 - acc: 0.0600

2021-09-07 05:06:10.220855: I tensorflow/core/profiler/lib/profiler_session.cc:131] Profiler session initializing.
2021-09-07 05:06:10.220884: I tensorflow/core/profiler/lib/profiler_session.cc:146] Profiler session started.
2021-09-07 05:06:10.220911: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1666] function cupti_interface_->Subscribe( &subscriber_, (CUpti_CallbackFunc)ApiCallback, this)failed with error CUPTI could not be loaded or symbol could not be found.

  2/753 [..............................] - ETA: 8:22 - loss: 5.6169 - acc: 0.1800 

2021-09-07 05:06:10.875966: I tensorflow/core/profiler/lib/profiler_session.cc:66] Profiler session collecting data.
2021-09-07 05:06:10.876019: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1757] function cupti_interface_->Finalize()failed with error CUPTI could not be loaded or symbol could not be found.
2021-09-07 05:06:10.888189: I tensorflow/core/profiler/internal/gpu/cupti_collector.cc:673]  GpuTracer has collected 0 callback api events and 0 activity events. 
2021-09-07 05:06:10.888916: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session tear down.
2021-09-07 05:06:10.890290: I tensorflow/core/profiler/rpc/client/save_profile.cc:136] Creating directory: /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10

2021-09-07 05:06:10.891045: I tensorflow/core/profiler/rpc/client/save_profile.cc:142] Dumped gzipped tool data for trace.json.gz to /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10/pop-os.trace.json.gz
2021-09-07 05:06:10.893450: I tensorflow/core/profiler/rpc/client/save_profile.cc:136] Creating directory: /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10

2021-09-07 05:06:10.893533: I tensorflow/core/profiler/rpc/client/save_profile.cc:142] Dumped gzipped tool data for memory_profile.json.gz to /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10/pop-os.memory_profile.json.gz
2021-09-07 05:06:10.893799: I tensorflow/core/profiler/rpc/client/capture_profile.cc:251] Creating directory: /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10
Dumped tool data for xplane.pb to /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10/pop-os.xplane.pb
Dumped tool data for overview_page.pb to /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10/pop-os.overview_page.pb
Dumped tool data for input_pipeline.pb to /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10/pop-os.input_pipeline.pb
Dumped tool data for tensorflow_stats.pb to /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10/pop-os.tensorflow_stats.pb
Dumped tool data for kernel_stats.pb to /home/mateen/Documents/ASR_Proj_1/Speech-Recognition-master/results/logs/train/plugins/profile/2021_09_07_05_06_10/pop-os.kernel_stats.pb

753/753 [==============================] - 632s 839ms/step - loss: 2.4952 - acc: 0.0527 - val_loss: 2.4861 - val_acc: 0.0340
Epoch 2/20
753/753 [==============================] - 613s 815ms/step - loss: 2.4812 - acc: 0.0441 - val_loss: 2.4860 - val_acc: 0.0340
Epoch 3/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4816 - acc: 0.0347 - val_loss: 2.4860 - val_acc: 0.0340
Epoch 4/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4819 - acc: 0.0348 - val_loss: 2.4857 - val_acc: 0.0340
Epoch 5/20
753/753 [==============================] - 605s 803ms/step - loss: 2.4815 - acc: 0.0347 - val_loss: 2.4854 - val_acc: 0.0340
Epoch 6/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4808 - acc: 0.0347 - val_loss: 2.4850 - val_acc: 0.0340
Epoch 7/20
753/753 [==============================] - 605s 803ms/step - loss: 2.4808 - acc: 0.0348 - val_loss: 2.4846 - val_acc: 0.0340
Epoch 8/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4812 - acc: 0.0347 - val_loss: 2.4843 - val_acc: 0.0340
Epoch 9/20
753/753 [==============================] - 605s 803ms/step - loss: 2.4809 - acc: 0.0347 - val_loss: 2.4840 - val_acc: 0.0340
Epoch 10/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4815 - acc: 0.0347 - val_loss: 2.4838 - val_acc: 0.0340
Epoch 11/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4812 - acc: 0.0347 - val_loss: 2.4835 - val_acc: 0.0340
Epoch 12/20
753/753 [==============================] - 606s 805ms/step - loss: 2.4811 - acc: 0.0347 - val_loss: 2.4833 - val_acc: 0.0340
Epoch 13/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4816 - acc: 0.0347 - val_loss: 2.4830 - val_acc: 0.0340
Epoch 14/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4807 - acc: 0.0347 - val_loss: 2.4827 - val_acc: 0.0340
Epoch 15/20
753/753 [==============================] - 605s 803ms/step - loss: 2.4810 - acc: 0.0347 - val_loss: 2.4825 - val_acc: 0.0340
Epoch 16/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4809 - acc: 0.0347 - val_loss: 2.4824 - val_acc: 0.0340
Epoch 17/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4814 - acc: 0.0347 - val_loss: 2.4823 - val_acc: 0.0340
Epoch 18/20
753/753 [==============================] - 606s 805ms/step - loss: 2.4812 - acc: 0.0347 - val_loss: 2.4823 - val_acc: 0.0340
Epoch 19/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4812 - acc: 0.0347 - val_loss: 2.4822 - val_acc: 0.0340
Epoch 20/20
753/753 [==============================] - 605s 804ms/step - loss: 2.4815 - acc: 0.0347 - val_loss: 2.4823 - val_acc: 0.0340
10766/10766 [==============================] - 95s 9ms/step
Test set accuracy: 3%

![index](https://user-images.githubusercontent.com/90254980/132337298-5a4e9bfd-d450-4395-a179-6708c7df9730.png)

**Comment by mhagglun** [comment]
Hey!
Sorry for the tardy response, I've been quite busy lately.
Anyways, regarding your problems:
1. I've not worked too much with Colab so you'll have to google yourself to an answer of how to make the notebook kernel to recognize the GPU. I have been using my own GPU when working on my DL models.

2.  This is likely due to the choice of parameters that you have set for training the model. Poorly chosen parameters will cause the model to not learn efficiently and it will not converge. I just pulled the code and ran it for only one epoch with the default parameters that I chose and the model did indeed show a convergence:

```
Model: "CNN"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
log_melgram_layer (LogMelgra (None, 27, 26, 1)         0         
_________________________________________________________________
conv2d (Conv2D)              (None, 25, 24, 16)        160       
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 24, 23, 32)        2080      
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 23, 22, 64)        8256      
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 22, 21, 128)       32896     
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 11, 10, 128)       0         
_________________________________________________________________
flatten (Flatten)            (None, 14080)             0         
_________________________________________________________________
dropout (Dropout)            (None, 14080)             0         
_________________________________________________________________
dense (Dense)                (None, 64)                901184    
_________________________________________________________________
dense_1 (Dense)              (None, 12)                780       
=================================================================
Total params: 945,356
Trainable params: 945,356
Non-trainable params: 0
_________________________________________________________________

Epoch 1/20  1/763 [..............................] - ETA: 19:21 - loss: 3.3145 - acc: 0.0300

2021-09-14 12:40:48.317122: I tensorflow/core/profiler/lib/profiler_session.cc:131] Profiler session initializing.
2021-09-14 12:40:48.317150: I tensorflow/core/profiler/lib/profiler_session.cc:146] Profiler session started.

  2/763 [..............................] - ETA: 12:43 - loss: 3.6287 - acc: 0.0150

2021-09-14 12:40:49.300828: I tensorflow/core/profiler/lib/profiler_session.cc:66] Profiler session collecting data.
2021-09-14 12:40:49.303407: I tensorflow/core/profiler/internal/gpu/cupti_collector.cc:673]  GpuTracer has collected 0 callback api events and 0 activity events. 
2021-09-14 12:40:49.305192: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session tear down.
2021-09-14 12:40:49.312653: I tensorflow/core/profiler/rpc/client/save_profile.cc:136] Creating directory: /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49

2021-09-14 12:40:49.313512: I tensorflow/core/profiler/rpc/client/save_profile.cc:142] Dumped gzipped tool data for trace.json.gz to /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49/oreo.trace.json.gz
2021-09-14 12:40:49.319510: I tensorflow/core/profiler/rpc/client/save_profile.cc:136] Creating directory: /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49

2021-09-14 12:40:49.319612: I tensorflow/core/profiler/rpc/client/save_profile.cc:142] Dumped gzipped tool data for memory_profile.json.gz to /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49/oreo.memory_profile.json.gz
2021-09-14 12:40:49.319827: I tensorflow/core/profiler/rpc/client/capture_profile.cc:251] Creating directory: /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49
Dumped tool data for xplane.pb to /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49/oreo.xplane.pb
Dumped tool data for overview_page.pb to /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49/oreo.overview_page.pb
Dumped tool data for input_pipeline.pb to /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49/oreo.input_pipeline.pb
Dumped tool data for tensorflow_stats.pb to /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49/oreo.tensorflow_stats.pb
Dumped tool data for kernel_stats.pb to /home/marcus/Desktop/Speech-Recognition/results/logs/train/plugins/profile/2021_09_14_12_40_49/oreo.kernel_stats.pb

763/763 [==============================] - ETA: 0s - loss: 1.0525 - acc: 0.4917
```

3. Because the model did not converge in your attempt, it is no wonder that it did so poorly during the testing. You can think of it as since it didn't converge during training, it has not learned any distinguishing features in the audio samples and so it pretty much defaults to doing the same guess for all samples.

I would advise you to try and train the model again with the parameters that I chose (or reasonably similar ones) for a few epochs and you should see a pretty clear convergence. 

Try to get the situation with Colab sorted out so that the runs will be quicker as well :smile: Good luck and let me know if you run into any additional problems.

**Comment by mhagglun** [comment]
Closing because of inactivity. Should any other issue arise feel free to re-open the issue.
